src/RestClient/Authenticators/OAuth/src/TokenSource/RefreshTokenSource.cs

```diff
@@ -1,35 +1,50 @@
 ﻿namespace ClickView.Extensions.RestClient.Authenticators.OAuth.TokenSource
 {
+    using Microsoft.Extensions.Logging;
+    using System;
     using System.Collections.Generic;
     using System.Threading;
     using System.Threading.Tasks;
-    using Microsoft.Extensions.Logging;
     using Tokens;
     using TokenStore;
+    using Utilities.Threading;
 
     public class RefreshTokenSource : ITokenSource
     {
         private readonly ILogger<RefreshTokenSource> _logger;
         private readonly ITokenStore _tokenStore;
         private readonly TokenClient _tokenClient;
+        private readonly TaskSingle<string> _singleTask;
 
         public RefreshTokenSource(ITokenStore tokenStore, TokenClient tokenClient, ILoggerFactory loggerFactory)
         {
             _tokenStore = tokenStore;
             _tokenClient = tokenClient;
             _logger = loggerFactory.CreateLogger<RefreshTokenSource>();
+
+            _singleTask = new TaskSingle<string>();
         }
 
         public async Task<IReadOnlyCollection<Token>> GetTokensAsync(CancellationToken cancellationToken = default)
         {
-            var refreshToken = await _tokenStore.GetTokenAsync(TokenType.RefreshToken).ConfigureAwait(false);
+            var refreshToken = await _tokenStore.GetTokenAsync(TokenType.RefreshToken);
             if (refreshToken == null)
                 return new List<Token>();
 
+            var accessToken = await _tokenStore.GetTokenAsync(TokenType.AccessToken);
+            if (accessToken?.ExpireTime != null && accessToken.ExpireTime > DateTimeOffset.UtcNow)
+                return new List<Token>
+                {
+                    accessToken,
+                    refreshToken
+                };
+
             _logger.LogDebug("Refreshing token");
 
-            var refreshTokenResponse = await _tokenClient.GetRefreshTokenAsync(refreshToken.Value, cancellationToken)
-                .ConfigureAwait(false);
+            var taskKey = "rt:" + refreshToken.Value;
```

Comment:
Since the TaskSingle is local, we dont really need to prefix it with a key since its only used here (saves a string concat allocation)

Reply:
done.